internal/mining/mining.go

```diff
@@ -1526,48 +1596,53 @@ mempoolLoop:
 			}
 		}
 
+		if miningView.IsRejected(prioItem.tx.Hash()) {
+			// If the transaction or any of its ancestors have been rejected,
+			// discard the transaction.
+			continue
+		}
+
+		ancestors := miningView.Ancestors(tx.Hash())
+		ancestorStats, _ := miningView.AncestorStats(tx.Hash())
+
+		prioItem.feePerKB = calcFeePerKb(prioItem.txDesc, ancestorStats)
+
 		// Enforce maximum block size.  Also check for overflow.
 		txSize := uint32(tx.MsgTx().SerializeSize())
-		blockPlusTxSize := blockSize + txSize
+		blockPlusTxSize := blockSize + txSize + uint32(ancestorStats.SizeBytes)
 		if blockPlusTxSize < blockSize ||
 			blockPlusTxSize >= g.policy.BlockMaxSize {
 			log.Tracef("Skipping tx %s (size %v) because it "+
 				"would exceed the max block size; cur block "+
 				"size %v, cur num tx %v", tx.Hash(), txSize,
 				blockSize, len(blockTxns))
 			logSkippedDeps(tx, deps)
+			miningView.Reject(tx.Hash())
 			continue
 		}
 
 		// Enforce maximum signature operations per block.  Also check
 		// for overflow.
-		numSigOps := int64(blockchain.CountSigOps(tx, false, isSSGen,
-			isTreasuryEnabled))
-		if blockSigOps+numSigOps < blockSigOps ||
-			blockSigOps+numSigOps > blockchain.MaxSigOpsPerBlock {
+		numSigOps := int64(prioItem.txDesc.NumSigOps) + ancestorStats.NumSigOps
+		numSigOpsBundle := numSigOps + ancestorStats.NumSigOps
+		if blockSigOps+numSigOpsBundle < blockSigOps ||
+			blockSigOps+numSigOpsBundle > blockchain.MaxSigOpsPerBlock {
 			log.Tracef("Skipping tx %s because it would "+
 				"exceed the maximum sigops per block", tx.Hash())
 			logSkippedDeps(tx, deps)
+			miningView.Reject(tx.Hash())
 			continue
 		}
 
-		// This isn't very expensive, but we do this check a number of times.
-		// Consider caching this in the mempool in the future. - Decred
-		numP2SHSigOps, err := blockchain.CountP2SHSigOps(tx, false,
-			isSSGen, blockUtxos, isTreasuryEnabled)
-		if err != nil {
-			log.Tracef("Skipping tx %s due to error in "+
-				"CountP2SHSigOps: %v", tx.Hash(), err)
-			logSkippedDeps(tx, deps)
-			continue
-		}
-		numSigOps += int64(numP2SHSigOps)
-		if blockSigOps+numSigOps < blockSigOps ||
-			blockSigOps+numSigOps > blockchain.MaxSigOpsPerBlock {
+		numSigOpsBundle += numSigOps + int64(prioItem.txDesc.NumP2SHSigOps) +
+			ancestorStats.NumP2SHSigOps
```

Comment:
Also unsure why `numSigOps` is being added here again when its already included in `numSigOpsBundle`

Reply:
Updated